adapters/yieldlab/yieldlab.go

```diff
@@ -0,0 +1,317 @@
+	}
+
+	if req.Device != nil {
+		q.Set("yl_rtb_ifa", req.Device.IFA)
+		q.Set("yl_rtb_devicetype", fmt.Sprintf("%v", req.Device.DeviceType))
+
+		if req.Device.ConnectionType != nil {
+			q.Set("yl_rtb_connectiontype", fmt.Sprintf("%v", req.Device.ConnectionType.Val()))
+		}
+
+		if req.Device.Geo != nil {
+			q.Set("lat", fmt.Sprintf("%v", req.Device.Geo.Lat))
+			q.Set("lon", fmt.Sprintf("%v", req.Device.Geo.Lon))
+		}
+	}
+
+	if req.App != nil {
+		q.Set("pubappname", req.App.Name)
+		q.Set("pubbundlename", req.App.Bundle)
+	}
+
+	gdpr, consent, err := a.getGDPR(req)
+	if err != nil {
+		return "", err
+	}
+	if gdpr != "" && consent != "" {
+		q.Set("gdpr", gdpr)
+		q.Set("consent", consent)
+	}
+
+	uri.RawQuery = q.Encode()
+
+	return uri.String(), nil
+}
+
+func (a *YieldlabAdapter) getGDPR(request *openrtb.BidRequest) (string, string, error) {
+	gdpr := ""
+	var extRegs openrtb_ext.ExtRegs
+	if request.Regs != nil {
+		if err := json.Unmarshal(request.Regs.Ext, &extRegs); err != nil {
+			return "", "", fmt.Errorf("failed to parse ExtRegs in Yieldlab GDPR check: %v", err)
+		}
+		if extRegs.GDPR != nil && (*extRegs.GDPR == 0 || *extRegs.GDPR == 1) {
+			gdpr = strconv.Itoa(int(*extRegs.GDPR))
+		}
+	}
+
+	consent := ""
+	if request.User != nil && request.User.Ext != nil {
+		var extUser openrtb_ext.ExtUser
+		if err := json.Unmarshal(request.User.Ext, &extUser); err != nil {
+			return "", "", fmt.Errorf("failed to parse ExtUser in Yieldlab GDPR check: %v", err)
+		}
+		consent = extUser.Consent
+	}
+
+	return gdpr, consent, nil
+}
+
+func (a *YieldlabAdapter) makeTargetingValues(params *openrtb_ext.ExtImpYieldlab) string {
+	values := url.Values{}
+	for k, v := range params.Targeting {
+		values.Set(k, v)
+	}
+	return values.Encode()
+}
+
+func (a *YieldlabAdapter) MakeRequests(request *openrtb.BidRequest, _ *adapters.ExtraRequestInfo) ([]*adapters.RequestData, []error) {
+	if len(request.Imp) == 0 {
+		return nil, []error{fmt.Errorf("invalid request %+v, no Impressions given", request)}
+	}
+
+	params, err := a.parseRequest(request)
+	if err != nil {
+		return nil, []error{err}
+	}
+
+	bidURL, err := a.makeEndpointURL(request, a.mergeParams(params))
+	if err != nil {
+		return nil, []error{err}
+	}
+
+	headers := http.Header{}
+	headers.Add("Accept", "application/json")
+	if request.Site != nil {
+		headers.Add("Referer", request.Site.Page)
+	}
+	if request.Device != nil {
+		headers.Add("User-Agent", request.Device.UA)
+		headers.Add("X-Forwarded-For", request.Device.IP)
+	}
+	if request.User != nil {
+		headers.Add("Cookie", "id="+request.User.BuyerUID)
+	}
+
+	return []*adapters.RequestData{{
+		Method:  "GET",
+		Uri:     bidURL,
+		Headers: headers,
+	}}, nil
+}
+
+// parseRequest extracts the Yieldlab request information from the request
+func (a *YieldlabAdapter) parseRequest(request *openrtb.BidRequest) ([]*openrtb_ext.ExtImpYieldlab, error) {
+	params := make([]*openrtb_ext.ExtImpYieldlab, 0)
```

Comment:
Would it be better to return an slice of `error`s instead?

Reply:
I've removed the error return completely as it's obsolete now.

Reply:
Sorry if I don't get that right, but didn't you suggest the exact opposite in this comment? https://github.com/prebid/prebid-server/pull/1287#discussion_r420500746

Reply:
Hi @aklinkert I'm sorry for the confusion. I think `parseRequest` with both with the `continue` statements and the `[]error` slice we are able to:
```
             IN                          OUT
[]Imp = {badExt, badExt, badExt}    --> exit and print error messages
[]Imp = {GoodExt, badExt, badExt}   --> process valid imps
[]Imp = {GoodExt, GoodExt, GoodExt} --> process valid imps
```
But you know better the context of your adapter so I'm ready to approve even if you leave it as is. Please let me know.

Reply:
@guscarreon, it‘s fine how it is, please merge :)